File: app/src/main/java/com/wolcano/musicplayer/music/ui/dialog/Dialogs.java
package com.wolcano.musicplayer.music.ui.dialog;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import android.text.Html;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.widget.Toast;

import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.Theme;
import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;
import com.wolcano.musicplayer.music.R;
import com.wolcano.musicplayer.music.content.Share;
import com.wolcano.musicplayer.music.mvp.models.Copy;
import com.wolcano.musicplayer.music.mvp.models.SongOnline;
import com.wolcano.musicplayer.music.mvp.models.Playlist;
import com.wolcano.musicplayer.music.mvp.models.Song;
import com.wolcano.musicplayer.music.ui.adapter.other.CopyItemAdapter;
import com.wolcano.musicplayer.music.ui.adapter.other.LikeItemAdapter;
import com.wolcano.musicplayer.music.ui.adapter.other.ShareItemAdapter;
import com.wolcano.musicplayer.music.utils.SongUtils;
import com.wolcano.musicplayer.music.utils.Utils;
import java.util.ArrayList;
import java.util.List;

public class Dialogs {

    private static Toast toast;

    public static void copyDialog(Context context, Song song) {
        ArrayList<Copy> copylist = new ArrayList<>();
        if (!song.getTitle().equals("<unknown>") && !song.getTitle().equals(""))
            copylist.add(new Copy(song.getTitle(), 0));
        if (!song.getArtist().equals("<unknown>") && !song.getArtist().equals(""))
            copylist.add(new Copy(song.getArtist(), 1));
        if (song.getAlbum() != null) {
            if (!song.getAlbum().equals("<unknown>") && !song.getAlbum().equals(""))
                copylist.add(new Copy(song.getAlbum(), 2));
        }
        final CopyItemAdapter adapter = new CopyItemAdapter(context, copylist);
        MaterialDialog dialog = new MaterialDialog.Builder(context)
                .title(R.string.copy_question)
                .itemsColor(ContextCompat.getColor(context, R.color.grey0))
                .adapter(adapter, null)
                .itemsCallback((dialog1, view, which, text) -> {
                    String replacedStr = text.toString().replaceAll("\"", "");
                    ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText(context.getString(R.string.copy_song_infos), replacedStr);
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(context, Html.fromHtml(context.getString(R.string.copy_to_clipboard, replacedStr)), Toast.LENGTH_SHORT).show();
                })
                .dividerColor(context.getResources().getColor(R.color.gray1))
                .theme(Theme.DARK)
                .show();
        adapter.setCallback(itemIndex -> showToastCopy(context, copylist.get(itemIndex).getText(), dialog));

    }
    public static void copyDialog(Context context,String album,String artist) {
        ArrayList<Copy> copylist = new ArrayList<>();

        if (!artist.equals("<unknown>") && !artist.equals(""))
            copylist.add(new Copy(artist, 1));
        if (album != null) {
            if (!album.equals("<unknown>") && !album.equals(""))
                copylist.add(new Copy(album, 2));
        }
        final CopyItemAdapter adapter = new CopyItemAdapter(context, copylist);
        MaterialDialog dialog = new MaterialDialog.Builder(context)
                .title(R.string.copy_question)
                .itemsColor(ContextCompat.getColor(context, R.color.grey0))
                .adapter(adapter, null)
                .itemsCallback((dialog1, view, which, text) -> {
                    String replacedStr = text.toString().replaceAll("\"", "");
                    ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText(context.getString(R.string.copy_song_infos), replacedStr);
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(context, Html.fromHtml(context.getString(R.string.copy_to_clipboard, replacedStr)), Toast.LENGTH_SHORT).show();
                })
                .dividerColor(context.getResources().getColor(R.color.gray1))
                .theme(Theme.DARK)
                .show();
        adapter.setCallback(itemIndex -> showToastCopy(context, copylist.get(itemIndex).getText(), dialog));

    }
    public static void copyDialog(Context context, SongOnline song) {
        ArrayList<Copy> copylist = new ArrayList<>();
        if (!song.getTitle().equals("<unknown>") && !song.getTitle().equals(""))
            copylist.add(new Copy(song.getTitle(), 0));
        if (!song.getArtistName().equals("<unknown>") && !song.getArtistName().equals(""))
            copylist.add(new Copy(song.getArtistName(), 1));
        final CopyItemAdapter adapter = new CopyItemAdapter(context, copylist);

        MaterialDialog dialog = new MaterialDialog.Builder(context)
                .title(R.string.copy_question)
                .itemsColor(ContextCompat.getColor(context, R.color.grey0))
                .adapter(adapter, null)
                .itemsCallback((dialog1, view, which, text) -> {
                    String replacedStr = text.toString().replaceAll("\"", "");
                    ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText(context.getString(R.string.copy_song_infos), replacedStr);
                    clipboard.setPrimaryClip(clip);
                    Toast.makeText(context, Html.fromHtml(context.getString(R.string.copy_to_clipboard, replacedStr)), Toast.LENGTH_SHORT).show();
                })
                .dividerColor(context.getResources().getColor(R.color.gray1))
                .theme(Theme.DARK)
                .show();
        adapter.setCallback(itemIndex -> showToastCopy(context, copylist.get(itemIndex).getText(), dialog));

    }

    private static void showToastCopy(Context context, String message, MaterialDialog dialog) {
        if (toast != null) {
            toast.cancel();
            toast = null;
        }
        ClipboardManager clipboard = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText(context.getString(R.string.copy_song_infos), message);
        clipboard.setPrimaryClip(clip);
        toast = Toast.makeText(context, Html.fromHtml(context.getString(R.string.copy_to_clipboard, message)), Toast.LENGTH_SHORT);
        toast.show();
        dialog.dismiss();
    }

    private static void showToastShare(Context context, String message, MaterialDialog dialog, Song song, int itemIndex) {
        if (itemIndex == 0) {
            Share.shareSong(context, Share.TYPE_SONG, song.getSongId());
        } else {
            Intent sharingIntent = new Intent(Intent.ACTION_SEND);
            sharingIntent.setType("text/plain");
            sharingIntent.putExtra(Intent.EXTRA_TEXT, message);
            context.startActivity(Intent.createChooser(sharingIntent, context.getString(R.string.share_name_ofsong)));
        }
        dialog.dismiss();
    }

    private static void showToastLike(Context context, MaterialDialog dialog, int itemIndex) {
        if (itemIndex == 0) {
            Utils.rateWolcano(context);
        } else {
            Utils.shareWolcano(context);
        }
        dialog.dismiss();
    }

    public static void shareDialog(Context context, Song song, Boolean isModel1) {
        CharSequence secondItem;
        if (song.getArtist().equals("<unknown>")) {
            secondItem = Html.fromHtml(context.getString(R.string.share_song_info_plain, song.getTitle()));
        } else {
            secondItem = Html.fromHtml(context.getString(R.string.share_song_info_wartist, song.getTitle(), song.getArtist()));
        }

        ArrayList<Copy> shareListModel1 = new ArrayList<>();
        shareListModel1.add(new Copy(context.getString(R.string.share_audio_file), 0));
        final ShareItemAdapter adapterModel1 = new ShareItemAdapter(context, shareListModel1);

        ArrayList<Copy> shareListModel0 = new ArrayList<>();
        shareListModel0.add(new Copy(context.getString(R.string.share_audio_file), 0));
        shareListModel0.add(new Copy(secondItem.toString(), 1));
        final ShareItemAdapter adapterModel0 = new ShareItemAdapter(context, shareListModel0);


        if (isModel1) {
            MaterialDialog dialogModel1 = new MaterialDialog.Builder(context)
                    .title(R.string.sharequestion)
                    .adapter(adapterModel1, null)
                    .theme(Theme.DARK)
                    .show();
            adapterModel1.setCallback(itemIndex -> showToastShare(context, shareListModel1.get(itemIndex).getText(), dialogModel1, song, itemIndex));
        } else {
            MaterialDialog dialogLocal = new MaterialDialog.Builder(context)
                    .title(R.string.sharequestion)
                    .adapter(adapterModel0, null)
                    .theme(Theme.DARK)
                    .show();
            adapterModel0.setCallback(itemIndex -> showToastShare(context, shareListModel0.get(itemIndex).getText(), dialogLocal, song, itemIndex));
        }

    }

    public static void addPlaylistDialog(final Context context, final Song song, List<Playlist> playlistList) {
        final CharSequence[] chars = new CharSequence[playlistList.size() + 1];
        chars[0] = context.getResources().getString(R.string.create_new_playlist);
        for (int i = 0; i < playlistList.size(); i++) {
            chars[i + 1] = playlistList.get(i).getName();
        }
        CharSequence title, artist;
        title = song.getTitle();
        artist = song.getArtist();
        Uri albumUri = Uri.parse("content://media/external/audio/media/" + song.getSongId() + "/albumart");

        Picasso.get().load(albumUri).into(new Target() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                update(bitmap);
            }

            @Override
            public void onBitmapFailed(Exception e, Drawable errorDrawable) {
                update(null);
            }

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {
            }

            private void update(Bitmap bitmap) {
                if (bitmap == null) {
                    bitmap = BitmapFactory.decodeResource(context.getResources(), R.drawable.album_art);
                }
                Drawable albumart = new BitmapDrawable(context.getResources(), bitmap);
                String wholeStr = title + "\n" + artist;
                SpannableString spanTitle = new SpannableString(wholeStr);
                spanTitle.setSpan(new ForegroundColorSpan(context.getResources().getColor(R.color.grey0)), (title + "\n").length(), wholeStr.length(), 0);

                new MaterialDialog.Builder(context)
                        .title(spanTitle)
                        .items(chars)
                        .icon(albumart)
                        .limitIconToDefaultSize()
                        .itemsCallback((dialog, itemView, which, text) -> {
                            if (which == 0) {
                                createPlaylistDialog(context, song.getSongId(), song.getTitle());
                                return;
                            }
                            SongUtils.addToPlaylist(context, song.getSongId(), playlistList.get(which - 1).getId(), song.getTitle());
                            dialog.dismiss();

                        }).build().show();

            }
        });

    }

    private static void createPlaylistDialog(Context context, long songID, String musicTitle) {
        new MaterialDialog.Builder(context)
                .title(R.string.create_new_playlist)
                .positiveText(R.string.create)
                .negativeText(R.string.cancel)
                .positiveColor(Utils.getAccentColor(context))
                .negativeColor(Utils.getAccentColor(context))
                .input(context.getString(R.string.playlist_name), "", false, new MaterialDialog.InputCallback() {
                    @Override
                    public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                        long playistId = SongUtils.createPlaylist(context, input.toString());

                        if (playistId != -1) {
                            SongUtils.addToPlaylist(context, songID, playistId, musicTitle);
                        } else {
                            Toast.makeText(context, R.string.create_playlist_fail, Toast.LENGTH_SHORT).show();
                        }
                    }
                }).build().show();
    }

    public static void likeDialog(Context context) {

        ArrayList<Copy> likeList = new ArrayList<>();
        likeList.add(new Copy(context.getString(R.string.pref_title_rate), 0));
        likeList.add(new Copy(context.getString(R.string.share_the_app), 1));
        final LikeItemAdapter adapterLike = new LikeItemAdapter(context, likeList);


        MaterialDialog dialog = new MaterialDialog.Builder(context)
                .title(R.string.like_dialog_title)
                .adapter(adapterLike, null)
                .theme(Theme.DARK)
                .show();
        adapterLike.setCallback(itemIndex -> showToastLike(context, dialog, itemIndex));

    }
}

File: app/src/main/java/com/wolcano/musicplayer/music/ui/dialog/Dialogs.kt
package com.wolcano.musicplayer.music.ui.dialog

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.net.Uri
import android.text.Html
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import android.widget.Toast
import androidx.core.content.ContextCompat
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.input.input
import com.afollestad.materialdialogs.list.customListAdapter
import com.afollestad.materialdialogs.list.listItems
import com.squareup.picasso.Picasso
import com.squareup.picasso.Picasso.LoadedFrom
import com.squareup.picasso.Target
import com.wolcano.musicplayer.music.R
import com.wolcano.musicplayer.music.content.Share
import com.wolcano.musicplayer.music.content.Share.shareSong
import com.wolcano.musicplayer.music.mvp.listener.ItemCallback
import com.wolcano.musicplayer.music.mvp.models.Copy
import com.wolcano.musicplayer.music.mvp.models.Playlist
import com.wolcano.musicplayer.music.mvp.models.Song
import com.wolcano.musicplayer.music.mvp.models.SongOnline
import com.wolcano.musicplayer.music.ui.adapter.customdialog.CopyItemAdapter
import com.wolcano.musicplayer.music.ui.adapter.customdialog.ShareItemAdapter
import com.wolcano.musicplayer.music.utils.SongUtils
import com.wolcano.musicplayer.music.utils.Utils
import java.util.*

object Dialogs {

    private var toast: Toast? = null

    fun copyDialog(context: Context, song: Song?) {
        val copylist = ArrayList<Copy>()
        if (song?.title != "<unknown>" && song?.title != "") copylist.add(Copy(song!!.title, 0))
        if (song.artist != "<unknown>" && song.artist != "") copylist.add(Copy(song.artist, 1))
        if (song.album != null) {
            if (song.album != "<unknown>" && song.album != "") copylist.add(Copy(song!!.album, 2))
        }
        val adapter = CopyItemAdapter(context, copylist)

        val dialog = MaterialDialog(context).show {
            title(R.string.copy_question)
            // itemsColor(ContextCompat.getColor(context, R.color.grey0))
            customListAdapter(adapter, null)
        }
        adapter.setCallback(object : ItemCallback {
            override fun onItemClicked(itemIndex: Int) {
                showToastCopy(
                    context,
                    copylist[itemIndex].text,
                    dialog
                )
            }
        })


    }

    fun copyDialog(context: Context, album: String?, artist: String) {
        val copylist = ArrayList<Copy>()
        if (artist != "<unknown>" && artist != "") copylist.add(Copy(artist, 1))
        if (album != null) {
            if (album != "<unknown>" && album != "") copylist.add(Copy(album, 2))
        }
        val adapter = CopyItemAdapter(context, copylist)
        val dialog = MaterialDialog(context).show {
            title(R.string.copy_question)
            // itemsColor(ContextCompat.getColor(context, R.color.grey0))
            customListAdapter(adapter, null)
        }
        adapter.setCallback(object : ItemCallback {
            override fun onItemClicked(itemIndex: Int) {
                showToastCopy(
                    context,
                    copylist[itemIndex].text,
                    dialog
                )
            }
        })
    }

    fun copyDialog(context: Context, song: SongOnline?) {
        val copylist = ArrayList<Copy>()
        if (song?.title != "<unknown>" && song?.title != "") copylist.add(Copy(song!!.title, 0))
        if (song.artistName != "<unknown>" && song?.artistName != "") copylist.add(
            Copy(
                song.artistName,
                1
            )
        )
        val adapter = CopyItemAdapter(context, copylist)
        val dialog = MaterialDialog(context).show {
            title(R.string.copy_question)
            // itemsColor(ContextCompat.getColor(context, R.color.grey0))
            customListAdapter(adapter, null)
        }
        adapter.setCallback(object : ItemCallback {
            override fun onItemClicked(itemIndex: Int) {
                showToastCopy(
                    context,
                    copylist[itemIndex].text,
                    dialog
                )
            }
        })
    }

    private fun showToastCopy(context: Context, message: String, dialog: MaterialDialog) {
        if (toast != null) {
            toast!!.cancel()
            toast = null
        }
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText(context.getString(R.string.copy_song_infos), message)
        clipboard.primaryClip = clip
        toast = Toast.makeText(
            context,
            Html.fromHtml(context.getString(R.string.copy_to_clipboard, message)),
            Toast.LENGTH_SHORT
        )
        toast?.show()
        dialog.dismiss()
    }


    private fun showToastShare(
        context: Context,
        message: String,
        dialog: MaterialDialog,
        song: Song,
        itemIndex: Int
    ) {
        if (itemIndex == 0) {
            shareSong(context, Share.TYPE_SONG, song.songId)
        } else {
            val sharingIntent = Intent(Intent.ACTION_SEND)
            sharingIntent.type = "text/plain"
            sharingIntent.putExtra(Intent.EXTRA_TEXT, message)
            context.startActivity(
                Intent.createChooser(
                    sharingIntent,
                    context.getString(R.string.share_name_ofsong)
                )
            )
        }
        dialog.dismiss()
    }

    private fun showToastLike(context: Context, dialog: MaterialDialog, itemIndex: Int) {
        if (itemIndex == 0) {
            Utils.rateWolcano(context)
        } else {
            Utils.shareWolcano(context)
        }
        dialog.dismiss()
    }

    fun shareDialog(context: Context, song: Song?, isModel1: Boolean) {
        val secondItem: CharSequence = if (song?.artist == "<unknown>") {
            Html.fromHtml(context.getString(R.string.share_song_info_plain, song.title))
        } else {
            Html.fromHtml(
                context.getString(
                    R.string.share_song_info_wartist,
                    song?.title,
                    song?.artist
                )
            )
        }
        val shareListModel1 = ArrayList<Copy>()
        shareListModel1.add(Copy(context.getString(R.string.share_audio_file), 0))
        val adapterModel1 = ShareItemAdapter(context, shareListModel1)
        val shareListModel0 = ArrayList<Copy>()
        shareListModel0.add(Copy(context.getString(R.string.share_audio_file), 0))
        shareListModel0.add(Copy(secondItem.toString(), 1))
        val adapterModel0 = ShareItemAdapter(context, shareListModel0)
        if (isModel1) {
            val dialogModel1 = MaterialDialog(context).show {
                title(R.string.sharequestion)
                customListAdapter(adapterModel1, null)
            }

            adapterModel1.setCallback(object : ItemCallback {
                override fun onItemClicked(itemIndex: Int) {
                    showToastShare(
                        context,
                        shareListModel1[itemIndex].text,
                        dialogModel1,
                        song!!,
                        itemIndex
                    )
                }
            })
        } else {
            val dialogLocal = MaterialDialog(context).show {
                title(R.string.sharequestion)
                customListAdapter(adapterModel0, null)
            }
            adapterModel0.setCallback(object : ItemCallback {
                override fun onItemClicked(itemIndex: Int) {
                    showToastShare(
                        context,
                        shareListModel0[itemIndex].text,
                        dialogLocal,
                        song!!,
                        itemIndex
                    )
                }
            })
        }
    }


    fun addPlaylistDialog(context: Context, song: Song?, playlistList: List<Playlist>) {
        val chars = mutableListOf<CharSequence>()
        chars.add(context.resources.getString(R.string.create_new_playlist))
        for (i in playlistList.indices) {
            chars.add(playlistList[i].name)
        }
        val title: CharSequence
        val artist: CharSequence
        title = song!!.title
        artist = song.artist
        val albumUri =
            Uri.parse("content://media/external/audio/media/" + song.songId + "/albumart")
        Picasso.get().load(albumUri).into(object : Target {
            override fun onBitmapLoaded(bitmap: Bitmap, from: LoadedFrom) {
                update(bitmap)
            }

            override fun onBitmapFailed(e: Exception, errorDrawable: Drawable) {
                update(null)
            }

            override fun onPrepareLoad(placeHolderDrawable: Drawable) {}
            private fun update(bitmap: Bitmap?) {
                var bitmap = bitmap
                if (bitmap == null) {
                    bitmap = BitmapFactory.decodeResource(context.resources, R.drawable.album_art)
                }
                val albumart: Drawable = BitmapDrawable(context.resources, bitmap)
                val wholeStr = """
                $title
                $artist
                """.trimIndent()
                val spanTitle = SpannableString(wholeStr)
                spanTitle.setSpan(
                    ForegroundColorSpan(ContextCompat.getColor(context, R.color.grey0)), """$title
""".length, wholeStr.length, 0
                )
                MaterialDialog(context).show {
                    title(text = spanTitle.toString())
                    listItems(items = chars) { dialog, index, _ ->
                        if (index == 0) {
                            createPlaylistDialog(
                                context,
                                song.songId,
                                song.title
                            )
                            return@listItems
                        }
                        SongUtils.addToPlaylist(
                            context,
                            song.songId,
                            playlistList[index - 1].id,
                            song.title
                        )
                        dialog.dismiss()

                    }
                    icon(drawable = albumart)
                }

            }

        })
    }

    private fun createPlaylistDialog(context: Context, songID: Long, musicTitle: String) {
        MaterialDialog(context).show {
            title(R.string.create_new_playlist)
            positiveButton(R.string.create)
            negativeButton(R.string.cancel)
            // positiveColor(Utils.getAccentColor(context))
            // negativeColor(Utils.getAccentColor(context))
            input(context.getString(R.string.playlist_name), prefill = "") { _, input ->
                val playlistId = SongUtils.createPlaylist(context, input.toString())
                if (playlistId != -1L) {
                    SongUtils.addToPlaylist(context, songID, playlistId, musicTitle)
                } else {
                    Toast.makeText(context, R.string.create_playlist_fail, Toast.LENGTH_SHORT)
                        .show()
                }
            }
        }

    }

}